staging/src/kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1/authorize_utils.go

```diff
@@ -0,0 +1,341 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+)
+
+func newCloneSourceHandler(dataVolume *DataVolume, dsGet dsGetFunc) (CloneSourceHandler, error) {
+	var pvcSource *DataVolumeSourcePVC
+	var snapshotSource *DataVolumeSourceSnapshot
+
+	if dataVolume.Spec.Source != nil {
+		if dataVolume.Spec.Source.PVC != nil {
+			pvcSource = dataVolume.Spec.Source.PVC
+		} else if dataVolume.Spec.Source.Snapshot != nil {
+			snapshotSource = dataVolume.Spec.Source.Snapshot
+		}
+	} else if dataVolume.Spec.SourceRef != nil && dataVolume.Spec.SourceRef.Kind == DataVolumeDataSource {
+		ns := dataVolume.Namespace
+		if dataVolume.Spec.SourceRef.Namespace != nil && *dataVolume.Spec.SourceRef.Namespace != "" {
+			ns = *dataVolume.Spec.SourceRef.Namespace
+		}
+		dataSource, err := dsGet(ns, dataVolume.Spec.SourceRef.Name)
+		if err != nil {
+			return CloneSourceHandler{}, err
+		}
+		if dataSource.Spec.Source.PVC != nil {
+			pvcSource = dataSource.Spec.Source.PVC
+		} else if dataSource.Spec.Source.Snapshot != nil {
+			snapshotSource = dataSource.Spec.Source.Snapshot
+		}
+	}
+
+	switch {
+	case pvcSource != nil:
+		return CloneSourceHandler{
+			CloneType:         pvcClone,
+			TokenResource:     tokenResourcePvc,
+			UserCloneAuthFunc: CanUserClonePVC,
+			SACloneAuthFunc:   CanServiceAccountClonePVC,
+			SourceName:        pvcSource.Name,
+			SourceNamespace:   pvcSource.Namespace,
+		}, nil
+	case snapshotSource != nil:
+		return CloneSourceHandler{
+			CloneType:         snapshotClone,
+			TokenResource:     tokenResourceSnapshot,
+			UserCloneAuthFunc: CanUserCloneSnapshot,
+			SACloneAuthFunc:   CanServiceAccountCloneSnapshot,
+			SourceName:        snapshotSource.Name,
+			SourceNamespace:   snapshotSource.Namespace,
+		}, nil
+	default:
+		return CloneSourceHandler{
+			CloneType: noClone,
+		}, nil
+	}
+}
+
+var (
+	tokenResourcePvc = metav1.GroupVersionResource{
+		Group:    "",
+		Version:  "v1",
+		Resource: "persistentvolumeclaims",
+	}
+
+	tokenResourceSnapshot = metav1.GroupVersionResource{
+		Group:    "snapshot.storage.k8s.io",
+		Version:  "v1",
+		Resource: "volumesnapshots",
+	}
+)
+
+type cloneType int
+
+const (
+	noClone cloneType = iota
+	pvcClone
+	snapshotClone
+)
+
+// CloneSourceHandler is a helper around determining the
+// correct way of authorizing a particular DataVolume
+// +k8s:deepcopy-gen=false
+// +k8s:openapi-gen=false
+type CloneSourceHandler struct {
+	CloneType         cloneType
+	TokenResource     metav1.GroupVersionResource
+	UserCloneAuthFunc UserCloneAuthFunc
+	SACloneAuthFunc   ServiceAccountCloneAuthFunc
+	SourceName        string
+	SourceNamespace   string
+}
+
+type createSarFunc func(*authorization.SubjectAccessReview) (*authorization.SubjectAccessReview, error)
+type dsGetFunc func(string, string) (*DataSource, error)
+
+// AuthorizationHelperProxy proxies calls to APIs used for DV authorization
+type AuthorizationHelperProxy interface {
```

Comment:
Can't believe I just thought of this and I'd be remiss to not bring it up, what if our api just takes a `dynamic.Interface` instead of the proxy?

PRO:  
Less code for clients

CON: 
more code in api (translating to/from unstructured)
adding `client-go` to go.mod

Don't care too much either way just want to get your opinion

Reply:
Would have to look up some references but the immediate concern is:
Would we still be able to restrict access similarly to how we do with proxy?
I mean, with the proxy, external entities know that CDI doesn't have access to the client itself other than
the implemented Get/Create methods

Reply:
With the proxy, callers know exactly what calls we are making, and that may be nice, but ultimately there is RBAC in place to keep the SA from doing anything bad.  And obviously the code is available to audit and make sure we're not reading secrets or whatever.

Reply:
Yes my concern is that we get passed a general SA that is probably not very limited,
unless you are suggesting to somehow reduce the SA

Reply:
No I don't think it is practical to create a special client with reduced permissions just to pass to this api.  I also don't think passing the dynamic client is a security concern.

Reply:
> I also don't think passing the dynamic client is a security concern.

I thought that is the reason we went with proxy structs in the first place

Reply:
> > I also don't think passing the dynamic client is a security concern.
> 
> I thought the reason we went with proxy structs in the first place was to avoid passing a full-blown client

The issue is with the typed client here:  https://github.com/kubevirt/containerized-data-importer/blob/main/pkg/client/clientset/versioned/typed/core/v1beta1/datavolume.go#L40

That would cause a circular dependency

Reply:
Ahh right.

> I thought the reason we went with proxy structs in the first place was to avoid passing a full-blown client

We used the sarProxy prior to this PR though (and thus prior to the circular dep potential), was this^ not the main concern?

Anyway my (not 100% trustworthy) gut is telling me that whipping out a dynamic client
when we have a defined set of actions on objects we know about is not right, so I lean towards current impl

Reply:
All past decisions are not made with perfect information.  If I could go back in time I would probably use the dynamic client.  But if you like using this interface, I don't care enough to keep belaboring the discussion.